### PR TITLE
report: correctly display CLS in budget table

### LINKED ISSUE #11177: CLS with Performance Budgets (budget.json) is always 0 ms in Lighthouse's HTML Report
**Cumulative Layout Shift is always shown with 0 ms in Lighthouse's HTML reports budgets sections. Metrics section shows the real number (not as ms)**

I try to run lighthouse npm with --budget-path= and want to test CLS with budget too. 

The json output with --output="json" --output-path="test.json" shows the measurement and overBudget. It's fine:
```
{
  "metric": "cumulative-layout-shift",
  "label": "Cumulative Layout Shift",
  "measurement": 1.178453672077047,
  "overBudget": 1.078453672077047
},
```

But in the html report it's not working. The HTML reports metric section is fine, **but the budgets section just shows 0 ms. Always. To show ms is wrong too.** 

CLS

CLS is mentioned here https://github.com/GoogleChrome/lighthouse/blob/master/docs/performance-budgets.md#timing-budgets as timings, so I go with a budget.json like this

```
[
  {
    "path": "/*",
    "timings": [
      {
        "metric": "total-blocking-time",
        "budget": 300
      },
      {
        "metric": "largest-contentful-paint",
        "budget": 2500
      },
      {
        "metric": "max-potential-fid",
        "budget": 130
      },
      {
        "metric": "cumulative-layout-shift",
        "budget": 0.1
      }
    ]
  }
]
```

Tested e.g. with lighthouse https://www.dailymail.co.uk/home/index.html --budget-path=budget-good.json

### PR DESCRIPTION
**Pull request by Beytoven**
Screen Shot 2020-08-03 at 2 11 44 PM

Fixes #11177

### CODE REVIEW

**Comment by connorjclark** [comment]
Nice! So to be clear, the issue was that the granularity was hiding the over-budget value for CLS?

**Comment by Beytoven** [comment]
> Nice! So to be clear, the issue was that the granularity was hiding the over-budget value for CLS?

That's part of it. The bigger issue was that the column was defined as 'ms' in the header and so any value in the column gets treated as such. This includes the number being rounded to the nearest 10 by default.

**Comment by brendankenny** [comment]
There's a simpler way of doing this without special-casing budgets, and I apologize it's not documented better. I think this audit is meant to use it for the SpeedIndex budget [based on this comment](https://github.com/GoogleChrome/lighthouse/blob/40baa221b87ae44c3184f9e6587ef173590ef595/lighthouse-core/audits/timing-budget.js#L135-L136), but it doesn't appear to actually be doing it.

Table headings set the rendered types for the items in their column (e.g. `measurement` and `overBudget` set `itemType: ms` here), but each item can override that rendered type individually by using the object form of an item value instead of the primitive form. An example of this is `renderCode()`, which can be invoked for a table value:
- with [just plain text](https://github.com/GoogleChrome/lighthouse/blob/40baa221b87ae44c3184f9e6587ef173590ef595/lighthouse-core/report/html/renderer/details-renderer.js#L258-L261) if the column is set to `itemType: code` or
- by an item that's an object with `type: code`, so the heading is ignored and [the object is rendered as a `code` detail](https://github.com/GoogleChrome/lighthouse/blob/40baa221b87ae44c3184f9e6587ef173590ef595/lighthouse-core/report/html/renderer/details-renderer.js#L231-L233).

So basically the CLS item needs to be something like `{type: 'numeric', value, granularity: 0.01}`. Unfortunately, this object form of `numeric` isn't something `detailsRenderer` knows how to render yet, so you'll need to
- add a `NumericValue` type in `audit-details.d.ts` ([alongside the similar `CodeValue`, `LinkValue`](https://github.com/GoogleChrome/lighthouse/blob/40baa221b87ae44c3184f9e6587ef173590ef595/types/audit-details.d.ts#L180-L198), etc)
- add the `this._renderNumeric()` call to [`_renderTableValue()` here](https://github.com/GoogleChrome/lighthouse/blob/40baa221b87ae44c3184f9e6587ef173590ef595/lighthouse-core/report/html/renderer/details-renderer.js#L245-L246)

**Comment by connorjclark** [comment]
Won't this be a breaking change?

**Comment by brendankenny** [comment]
It'll be a new type, so it'll fall back to the less good unknown details type in old renderers, but it's a new feature, not a breaking change. It would be the same for a `BudgetValue`, too.

**Comment by connorjclark** [comment]
I meant for either the current change or just doing what you suggested: programmatic consumers of budgets likely aren't expecting `overBudget` to be an object.

**Comment by patrickhulce** [comment]
> Does this meet our threshold of a breaking change?

Given that it's not adding a new details renderer type, I would say probably not *requiring* a major version rev, but it definitely could break some integrations (the way Lighthouse CI processes resource budgets would break if we did this there) so worth giving a heads up to our big programmatic consumers. If it were close enough to 7.0 I would probably say it's worth waiting :)

I think we once threw around the idea of creating a locked issue that interested consumers could subscribe to for notable changes like this?

**Comment by brendankenny** [comment]
> Given that it's not adding a new details renderer type, I would say probably not _requiring_ a major version rev, but it definitely could break some integrations

We say `details` aren't generally part of our breaking changes, but there probably are exceptions for specific audits (budgets, metrics, etc) that people might reasonably rely on raw access staying stable.

OTOH, this is arguably a bug fix in line with the published details types, and people relying on the buggy behavior will have to adapt :)

Waiting if we think it's going to break someone or just relying on a strong notification of the fix both SGTM.

File: lighthouse-core/audits/timing-budget.js [hunk 1]
```diff
@@ -89,23 +89,47 @@ class TimingBudget extends Audit {
...
   static tableItems(budget, summary) {
+    /** @type {Array<BudgetItem>} */
+    const items = [];
     if (!budget.timings) {
-      return [];
+      return items;
     }
-    return budget.timings.map((timingBudget) => {
+
+    for (const timingBudget of budget.timings) {
```
Comment (connorjclark) [inline lighthouse-core/audits/timing-budget.js:98]: <3

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 2]
```diff
@@ -556,6 +559,7 @@ class DetailsRenderer {
     return element;
   }
 
+
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/details-renderer.js:562]:
```suggestion
```

File: lighthouse-core/audits/timing-budget.js [hunk 3]
```diff
@@ -89,23 +89,47 @@ class TimingBudget extends Audit {
...
-      return {
-        metric: metricName,
-        label,
-        measurement,
-        overBudget,
-      };
-    }).sort((a, b) => {
-      return (b.overBudget || 0) - (a.overBudget || 0);
+
+      if (metricName === 'cumulative-layout-shift') {
```
Comment (connorjclark) [inline lighthouse-core/audits/timing-budget.js:105]: It's unfortunate how complex this has made this fn + the sort at the end.

wb ~L167, where `tableItems` is called, we filter the results and replace the CLS measurement value instead ?
